src/neural/network_demux.cc

```diff
@@ -0,0 +1,224 @@
+    Abort();
+    Wait();
+    // Unstuck waiting computations.
+    while (!queuedemux_.empty()) {
+      queuedemux_.front()->NotifyComplete();
+      queuedemux_.pop();
+    }
+  }
+
+  void Worker() {
+    // While Abort() is not called (and it can only be called from destructor).
+    while (!abort_) {
+      {
+        {
+          std::unique_lock<std::mutex> lock(mutex_);
+          // Wait until there's come work to compute.
+          cv_.wait(lock, [&] { return abort_ || !queue_.empty(); });
+          if (abort_) break;
+        }
+
+        // While there is a work in queue, process it.
+        while (true) {
+          NetworkComputation* to_compute;
+          DemuxingComputation* to_notify;
+          {
+            std::unique_lock<std::mutex> lock(mutex_);
+            if (queue_.empty()) break;
+            to_compute = queue_.front();
+            queue_.pop();
+            to_notify = queuedemux_.front();
+            queuedemux_.pop();
+          }
+          to_compute->ComputeBlocking();
+          to_notify->NotifyComplete();
+        }
+      }
+    }
+  }
+
+  void Abort() {
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      abort_ = true;
+    }
+    cv_.notify_all();
+  }
+
+  void Wait() {
+    while (!threads_.empty()) {
+      threads_.back().join();
+      threads_.pop_back();
+    }
+  }
+
+  std::vector<std::unique_ptr<Network>> networks_;
+  std::queue<NetworkComputation*> queue_;
+  std::queue<DemuxingComputation*> queuedemux_;
+  int minimum_split_size_ = 0;
+  std::atomic<long long> counter_;
+  bool abort_ = false;
+
+  std::mutex mutex_;
+  std::condition_variable cv_;
+
+  std::vector<std::thread> threads_;
+};
+
+void DemuxingComputation::ComputeBlocking() {
+  if (GetBatchSize() == 0) return;
+  partial_size_ = (GetBatchSize() + network_->networks_.size() - 1) /
+                  network_->networks_.size();
+  if (partial_size_ < network_->minimum_split_size_) {
+    partial_size_ = std::min(GetBatchSize(), network_->minimum_split_size_);
+  }
+  int splits = (GetBatchSize() + partial_size_ - 1) / partial_size_;
+
+  std::unique_lock<std::mutex> lock(mutex_);
+  dataready_ = splits;
+  int cur_idx = 0;
+  for (int j=0; j < splits; j++) {
+    long long net_idx = ++(network_->counter_) % network_->networks_.size();
+    parents_.emplace_back(network_->networks_[net_idx]->NewComputation());
+    for (int i = cur_idx; i < std::min(GetBatchSize(), cur_idx + partial_size_);
+         i++) {
+      parents_.back()->AddInput(std::move(planes_[i]));
+    }
+    network_->Enqueue(parents_.back().get(), this);
+    cur_idx += partial_size_;
+  }
+  dataready_cv_.wait(lock, [this]() { return dataready_ == 0; });
+}
+
+std::unique_ptr<Network> MakeDemuxingNetwork(const WeightsFile& weights,
+                                             const OptionsDict& options) {
+  return std::make_unique<DemuxingNetwork>(weights, options);
+}
+
+REGISTER_NETWORK("demux", MakeDemuxingNetwork, -1000)
```

Comment:
Maybe some unique values instead of -1000 (-1001, -999?) so that their priority order is stable.